### PR TITLE
[MIN-101] Add readability score threshold

### PR DESCRIPTION
**Pull request by osw282**
This PR implements a readability score threshold detection. If user agrees to the data sharing consent, whenever a readability score is below a threshold, the score, question and response will automatically be sent to our metric service and store them in the metric database. Furthermore, a message below the response will also be displayed in streamlit when the score is below our set threshold.

A new table is created in the metric database `readability_threshold` for storing the data that triggered the readability threshold. The metric service is updated to send and query this table.

Minor refactoring is also done on the validate data functions in `metric_service.py` to reduce code duplication.

Tests are also updated.

### CODE REVIEW

File: app/app.py [hunk 1]
```diff
@@ -448,16 +482,88 @@ def show_thumbs(
...
+    """Create the readability score threshold collector.
+
+    This function will display a message saying that we have detected the readability score of the response is below a certain threshold.
+    If the user agree to share data with us, we will send this data to our metric service and store them in the metric database.
+
+    Args:
+        metric_service_endpoint (str): the metric service endpoint
+        readability_scores (Dict[str, Dict[str, Union[str, float]]]): a dictionary containing the score, question response.
+    """
+    readability_score_threshold = 55.0
```
Comment (JonoCX) [inline app/app.py]: Nit: I'd have this as a static variable up at the top of here (easier to configure)

**Review by JonoCX — APPROVED** [review]
LGTM! 🌯

File: app/app.py [hunk 2]
```diff
@@ -58,6 +58,8 @@
 
 CONVERSATIONAL_MEMORY_SIZE = 3
 
+READABILITY_SCORE_THRESHOLD = 55.0
```
Comment (dudeperf3ct) [inline app/app.py:61]: Should this be a streamlit slider or a fixed constant?

File: app/app.py [hunk 3]
```diff
@@ -396,50 +398,84 @@ def show_data_collection_permission() -> None:
...
-    metric_service_endpoint: str, user_rating: str, question: str, full_response: str
+def post_feedback_data_to_metric_service(
+    metric_service_endpoint: str, data: Dict[str, Union[str, Any]]
 ) -> None:
-    """Send the user feedback data to the metric service's `user_feedback` route using a POST request.
+    """Sends data to the metric service using a POST request.
 
+    Args:
+        metric_service_endpoint (str): The metric service endpoint where the data is expected.
+        data (dict): The data to be sent.
```
Comment (dudeperf3ct) [inline app/app.py]: Docstring mismatch

File: app/app.py [hunk 4]
```diff
@@ -396,50 +398,84 @@ def show_data_collection_permission() -> None:
...
+    readability_score: float,
+    question: str,
+    response: str,
+    dataset: str,
 ) -> None:
-    """This function construct two buttons using streamlit buttons and columns.
+    """Send the user question and response data to the metric service.
+    Args:
+        metric_service_endpoint (str): the metric service endpoint where the below readability score threshold data is expected
+        readability_score (str): the readability score
```
Comment (dudeperf3ct) [inline app/app.py]: Docstring mismatch

File: app/app.py [hunk 5]
```diff
@@ -448,16 +484,87 @@ def show_thumbs(
...
+    With columns, we can define where buttons should be places, in our case, two buttons side by side on the lower left.
+    Reference: https://discuss.streamlit.io/t/st-button-in-one-line/25966/2
+
+    A readability score threshold collector which display a message when the readability score falls below a threshold and store the score, question and response if user agree to share.
+
+    Args:
+        metric_service_endpoint (str): the metric service endpoint where the readability is computed
+        question (str): the question user asked
+        full_response (str): the full response the user rated.
+        readability_scores (Dict[str, float]): the readability score for Mind and NHS responses.
```
Comment (dudeperf3ct) [inline app/app.py]: Docstring mismatch

File: app/app.py [hunk 6]
```diff
@@ -573,6 +680,10 @@ def main() -> None:
                 with st.chat_message("assistant"):
                     full_response = "Here's what the NHS and Mind each have to say:\n\n"
                     message_placeholder = st.empty()
+                    feedback_placeholder = (
+                        st.empty()
+                    )  # Placeholder for the thumbs up and thumbs down button
```
Comment (dudeperf3ct) [inline app/app.py]: I think this should be fixed right? Is the develop branch not merged into this branch yet?

**Review by dudeperf3ct — CHANGES_REQUESTED** [review]
Mostly comments around docstring mismatches. Implementation looks good

File: app/app.py (hunk 2 above)
Comment (osw282) [reply app/app.py:61]: A fixed constant. User probably shouldn't be able to change this. 55 is what we came up with for now looking the chart [here](https://en.wikipedia.org/wiki/Flesch%E2%80%93Kincaid_readability_tests).

File: app/app.py (hunk 6 above)
Comment (osw282) [reply app/app.py]: Weird. It is merged. Fixed it again.

File: app/app.py [hunk 7]
```diff
@@ -396,50 +398,84 @@ def show_data_collection_permission() -> None:
...
-    """Send the user feedback data to the metric service's `user_feedback` route using a POST request.
+    """Sends data to the metric service using a POST request.
+
+    Args:
+        metric_service_endpoint (str): The metric service endpoint where the data is expected.
+        data (Dict[str, Union[str, Any]]): The data to be sent.
+    """
+    if (
+        st.session_state.data_sharing_consent == True
+    ):  # Store the response to metric database if user agrees to share.
```
Comment (dudeperf3ct) [inline app/app.py]: Can this inline comment be moved as well? It creates weird formatting.

File: app/app.py [hunk 8]
```diff
@@ -574,6 +681,12 @@ def main() -> None:
                     full_response = "Here's what the NHS and Mind each have to say:\n\n"
                     message_placeholder = st.empty()
 
+                    feedback_placeholder = (
+                        st.empty()
+                    )  # Placeholder for the thumbs up and thumbs down button
```
Comment (dudeperf3ct) [inline app/app.py]: Hmm, the inline comment is not fixed.

**Review by dudeperf3ct — CHANGES_REQUESTED** [review]

File: app/app.py (hunk 8 above)
Comment (osw282) [reply app/app.py]: Sorry about this, not sure what happened. Can confirm it's actually fixed now [line 689](https://github.com/fuzzylabs/MindGPT/pull/69/files#diff-fc3b1a787c6d94d379c141d039d3d5eb98ba8804720945de123abb50a167006cR689).

**Review by dudeperf3ct — APPROVED** [review]
Nice work @osw282

**Pull request merged**